numpy/lib/function_base.py

```diff
@@ -1901,26 +1901,182 @@ def diff(a, n=1, axis=-1):
     array([1, 1], dtype='timedelta64[D]')
 
     """
-    if n == 0:
-        return a
-    if n < 0:
-        raise ValueError(
-            "order must be non-negative but got " + repr(n))
+    def diff_func(a, b):
+        return not_equal(a, b) if a.dtype == np.bool_ else subtract(a, b)
```

Comment:
In case anyone asks, the decision on which function to use is made dynamically at each iteration instead of directly in `diff` itself to avoid problems when some magical non-bool dtype diffs to a bool somehow (like `datetime64` does to `timedelta64`). This may be unnecessarily pedantic, but I don't think it adds much of a processing burden in the `n>1` cases.